consensus/src/chained_bft/block_storage/block_tree.rs

```diff
@@ -193,6 +193,18 @@ where
     pub(super) fn insert_quorum_cert(&mut self, qc: QuorumCert) -> Result<(), BlockTreeError> {
         let block_id = qc.certified_block_id();
         let qc = Arc::new(qc);
+
+        // Safety invariant: For any two quorum certificates qc1, qc2 in the block store,
```

Comment:
I am not sure this invariant holds with the current equality operator. We don't do anything special for the '==' operator of QC, which means that we're checking all the signatures on `LedgerInfoWithSignatures` when running `==` on the QCs.
Both nil blocks and secondary proposers might create a situation with the same block being certified by different quorums: all the fields are the same but the sets of signatures used for certifying a block are different.
One way to check the thing you want to check here is to compare qc.ledger_info().hash()?

Reply:
> Both nil blocks and secondary proposers might create a situation with the same block being certified by different quorums: all the fields are the same but the sets of signatures used for certifying a block are different.

In the case where the same block is certified by different quorums, shouldn't the `round numbers` be different for both these quorums? My understanding is that the voting rules do not allow for more than one quorum to be formed in a given `round`, but I might be missing some implementation detail.

Reply:
Ok I take it back. I think there are two things to distinguish here: 1) a possibility of having block B certified by two identical QCs that have different signatures, and 2) a possibility of calling `insert_quorum_cert()` with these two different QCs.

1) There can be at most one block certified per round, this is a fundamental property we should have in the protocol. However, the signatures for the block certificate can be aggregated by different validators. When a validator votes for B it sends its vote for the primary and the secondary proposers of the next round. The primary and the secondary proposers aggregate the signatures and generate the QC for B independently. The two QCs are generated with identical information but with potentially different signatures that verify this information.
2) As far as I can follow our code I can't find a place in which we'd try to insert the QC for the same block twice to the store. Hence, the whole story of "same certificate with different signatures" should not matter much.

I would still prefer us to check something more intuitive: if we want to make sure that there are not two certified blocks per round, let's do it this way! (without checking the signatures).
Btw, did you figure out the reason for a failure we had in CircleCI in this assertion?

Reply:
The confusion here comes from the equality check. Two qc is equivalent if they certify the same block, but they may have different set of signatures. In this case, the current equality check will fail thus @dmitri-perelman suggested to compare the certified content(`qc.ledger_info().hash()`) instead.

>the voting rules do not allow for more than one quorum to be formed in a given round

Voting rule prevents more than one quorum for different blocks, it's possible to aggregate C(2f+1, 3f+1) different qc for the same block.

Reply:
Thanks for the clarification! 
> I would still prefer us to check something more intuitive: if we want to make sure that there are not two certified blocks per round, let's do it this way! (without checking the 
signatures).

I agree. I am leaving the current check as is.

> Btw, did you figure out the reason for a failure we had in CircleCI in this assertion?

The check was previously a `debug_checked_precondition` which was adding an assertion at runtime potentially causing a linear time traversal of the `block store`. This was causing a timeout in the validator (but wasn't reproducible locally). I have changed the checked to a `precondition` which won't add this runtime check and will only make sense when run with `MIRAI`(apart from providing documentation).